packages/fleet/lib/models/nodes.ts

```diff
@@ -0,0 +1,218 @@
+        cpu_milli: nodeProps.cpuMilli,
+        memory_mb: nodeProps.memoryMb,
+        storage_mb: nodeProps.storageMb,
+        error: null,
+        created_at: now,
+        last_state_transition_at: now
+    };
+    try {
+        const inserted = await db.from<DBNode>(NODES_TABLE).insert(newNode).returning('*');
+        if (!inserted?.[0]) {
+            return Err(new Error(`Error: no node '${nodeProps.routingId}' created`));
+        }
+        return Ok(DBNode.from(inserted[0]));
+    } catch (err: unknown) {
+        return Err(new Error(`Error creating node '${nodeProps.routingId}': ${stringifyError(err)}`));
+    }
+}
+export async function get(db: knex.Knex, nodeId: number, options: { forUpdate: boolean } = { forUpdate: false }): Promise<Result<Node>> {
+    const query = db.select<DBNode>('*').from(NODES_TABLE).where({ id: nodeId }).first();
+    if (options.forUpdate) {
+        query.forUpdate();
+    }
+    const node = await query;
+    if (!node) {
+        return Err(new Error(`Node with id '${nodeId}' not found`));
+    }
+    return Ok(DBNode.from(node));
+}
+
+export async function search(
+    db: knex.Knex,
+    params: {
+        states: [NodeState, ...NodeState[]]; // non-empty array
+        routingId?: RoutingId;
+        cursor?: number;
+        limit?: number;
+    }
+): Promise<
+    Result<{
+        nodes: Map<RoutingId, Node[]>;
+        nextCursor?: number;
+    }>
+> {
+    const limit = params.limit || 1000;
+    const query = db
+        .select<DBNode[]>('*')
+        .from(NODES_TABLE)
+        .whereIn('state', params.states)
+        .orderBy('id')
+        .limit(limit + 1); // fetch one more than limit to determine if there are more results
+
+    if (params.routingId) {
+        query.where({ routing_id: params.routingId });
+    }
+    if (params.cursor) {
+        query.where('id', '>=', params.cursor);
+    }
+
+    const nodes = await query;
+
+    const nextCursor = nodes.length > limit ? nodes.pop()?.id : undefined;
+
+    const nodesMap = new Map<RoutingId, Node[]>();
+
+    for (const node of nodes) {
+        const routingId = node.routing_id;
+        const existingNodes = nodesMap.get(routingId) || [];
+        existingNodes.push(DBNode.from(node));
+        nodesMap.set(routingId, existingNodes);
+    }
+
+    return Ok({
+        nodes: nodesMap,
+        ...(nextCursor ? { nextCursor } : {})
+    });
+}
+
+export async function transitionTo(
+    db: knex.Knex,
+    props: {
+        nodeId: number;
+        newState: Omit<NodeState, 'ERROR'>;
+    }
+): Promise<Result<Node>> {
+    return db.transaction(async (trx) => {
+        const node = await get(trx, props.nodeId, { forUpdate: true });
+        if (node.isErr()) {
+            return Err(new Error(`Node '${props.nodeId}' not found`));
+        }
+
+        const transition = NodeStateTransition.validate({ from: node.value.state, to: props.newState });
+        if (transition.isErr()) {
+            return Err(transition.error);
```

Comment:
```suggestion
            return transition.error;
```

Reply:
A result must be returned. I could maybe do `return transition.mapError(...)` but I don't think it is more readable